File: src/core/mcp/MCPCoordinator.ts
// src/core/mcp/MCPCoordinator.ts
// Implementation of the MCP Coordinator

import { Client } from '@modelcontextprotocol/sdk';
import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import { loadMCPConfig, MCPConfig, MCPServerConfig } from '../../config/mcpConfig';
import {
  MCPCoordinatorConfig,
  ServerToolDescriptionPrompts,
  ToolCallRequest,
  ToolCallResponse,
  ToolDefinition,
  ToolExecutionOptions,
  ToolRegistryEntry,
} from '../../types/mcp';
import {
  CircuitOpenError,
  ConfigurationError,
  ToolExecutionError,
  ToolNotFoundError,
  ToolValidationError,
} from '../../utils/errors';
import { createComponentLogger } from '../../utils/logger';
import { generateFallbackToolDescription } from './utils/generateFallbackPrompt';

const logger = createComponentLogger('MCPCoordinator');

/**
 * MCP Coordinator
 *
 * Manages connections to MCP servers, discovers tools, retrieves server-specific
 * tool usage guidance prompts, and handles tool execution requests.
 */
export class MCPCoordinator {
  private toolRegistry = new Map<string, ToolRegistryEntry>();
  private toolDescriptionPrompts = new Map<string, string>();
  private clients = new Map<string, Client>();
  private activeExecutions = new Map<
    string,
    {
      abortController: AbortController;
      timeout: NodeJS.Timeout;
    }
  >();
  private ajv: Ajv;
  private config: MCPCoordinatorConfig;

  /**
   * Creates a new MCP Coordinator
   * @param config Configuration options
   */
  constructor(config: Partial<MCPCoordinatorConfig> = {}) {
    this.config = {
      configPath: config.configPath || process.env.MCP_JSON_PATH || './mcp.json',
      circuitBreakerThreshold: config.circuitBreakerThreshold || 5,
      circuitResetTimeMs: config.circuitResetTimeMs || 60000, // 1 minute
      defaultTimeoutMs: config.defaultTimeoutMs || 10000, // 10 seconds
    };

    logger.info({ configPath: this.config.configPath }, 'Creating MCP Coordinator');

    // Initialize JSON Schema validator
    this.ajv = new Ajv({
      allErrors: true,
      strict: false,
    });
    addFormats(this.ajv);
  }

  /**
   * Initializes the MCP Coordinator
   *
   * This loads the configuration, connects to servers, discovers tools,
   * fetches "tool-descriptions" prompts, and builds the tool registry.
   */
  async initialize(): Promise<void> {
    logger.info('Initializing MCP Coordinator');

    try {
      // Load and validate the configuration
      const mcpConfig = loadMCPConfig(this.config.configPath);
      logger.info(
        { serverCount: Object.keys(mcpConfig.servers).length },
        'Loaded MCP configuration',
      );

      // Initialize clients and discover tools
      await this.initializeClients(mcpConfig);
      await this.discoverToolsAndDescriptions();

      logger.info(
        { toolCount: this.toolRegistry.size },
        'MCP Coordinator initialized successfully',
      );
    } catch (error) {
      logger.error({ error }, 'Failed to initialize MCP Coordinator');
      throw error;
    }
  }

  /**
   * Initializes clients for all servers in the configuration
   * @param mcpConfig The MCP configuration
   */
  private async initializeClients(mcpConfig: MCPConfig): Promise<void> {
    for (const [serverId, serverConfig] of Object.entries(mcpConfig.servers)) {
      try {
        logger.debug({ serverId, transport: serverConfig.transport }, 'Initializing MCP client');

        const client = await this.createClient(serverId, serverConfig);
        this.clients.set(serverId, client);

        logger.info({ serverId }, 'Successfully connected to MCP server');
      } catch (error) {
        logger.error({ serverId, error }, 'Failed to initialize MCP client');
        // Continue with other servers even if one fails
      }
    }
  }

  /**
   * Creates a client for an MCP server
   * @param serverId The server ID
   * @param serverConfig The server configuration
   * @returns An initialized MCP client
   */
  private async createClient(serverId: string, serverConfig: MCPServerConfig): Promise<Client> {
    try {
      let client: Client;

      if (serverConfig.transport === 'stdio') {
        if (!serverConfig.command) {
          throw new ConfigurationError(
            `Missing 'command' for stdio transport in server '${serverId}'`,
          );
        }

        client = new Client({
          transport: 'stdio',
          command: serverConfig.command,
          args: serverConfig.args || [],
        });
      } else if (serverConfig.transport === 'http') {
        if (!serverConfig.url) {
          throw new ConfigurationError(`Missing 'url' for http transport in server '${serverId}'`);
        }

        client = new Client({
          transport: 'http',
          url: serverConfig.url,
        });
      } else {
        throw new ConfigurationError(
          `Unsupported transport '${serverConfig.transport}' for server '${serverId}'`,
        );
      }

      // Connect to the server
      await client.connect();
      return client;
    } catch (error) {
      logger.error({ serverId, error }, 'Failed to create MCP client');
      throw new ConfigurationError(
        `Failed to create MCP client for server '${serverId}': ${(error as Error).message}`,
      );
    }
  }

  /**
   * Discovers tools and fetches/generates tool descriptions
   */
  private async discoverToolsAndDescriptions(): Promise<void> {
    for (const [serverId, client] of this.clients.entries()) {
      let toolDefinitions: ToolDefinition[] = [];

      try {
        // Discover tools
        logger.debug({ serverId }, 'Discovering tools');
        const toolsResponse = await client.listTools();

        if (!toolsResponse?.tools || !Array.isArray(toolsResponse.tools)) {
          logger.warn({ serverId, response: toolsResponse }, 'Invalid response from tools/list');
          continue;
        }

        toolDefinitions = toolsResponse.tools;
        logger.info({ serverId, toolCount: toolDefinitions.length }, 'Discovered tools');

        // Attempt to fetch "tool-descriptions" prompt
        try {
          logger.debug({ serverId }, 'Fetching tool-descriptions prompt');
          const promptResponse = await client.getPrompt('tool-descriptions');

          // Validate prompt response
          if (
            promptResponse?.messages &&
            Array.isArray(promptResponse.messages) &&
            promptResponse.messages.length === 1 &&
            promptResponse.messages[0].role === 'user' &&
            promptResponse.messages[0].content.type === 'text'
          ) {
            this.toolDescriptionPrompts.set(serverId, promptResponse.messages[0].content.text);
            logger.info({ serverId }, 'Successfully fetched tool-descriptions prompt');
          } else {
            throw new Error('Invalid prompt structure received');
          }
        } catch (promptError) {
          // Generate fallback description if prompt fetch fails
          logger.warn(
            { serverId, error: promptError },
            'Tool-descriptions prompt missing or invalid, generating fallback',
          );
          const fallbackPrompt = generateFallbackToolDescription(serverId, toolDefinitions);
          this.toolDescriptionPrompts.set(serverId, fallbackPrompt);
          logger.info({ serverId }, 'Generated fallback tool description prompt');
        }

        // Register tools in the registry
        this.registerTools(serverId, client, toolDefinitions);
      } catch (error) {
        logger.error({ serverId, error }, 'Failed to discover tools or fetch descriptions');
        // Continue with other servers even if one fails
      }
    }
  }

  /**
   * Registers tools in the tool registry
   * @param serverId The server ID
   * @param client The MCP client
   * @param tools The tool definitions
   */
  private registerTools(serverId: string, client: Client, tools: ToolDefinition[]): void {
    for (const tool of tools) {
      const qualifiedName = `${serverId}:${tool.name}`;

      // Check for name collisions with unqualified names
      const unqualifiedCollision = [...this.toolRegistry.values()].find(
        (entry) => entry.qualifiedName.endsWith(`:${tool.name}`) && entry.serverId !== serverId,
      );

      if (unqualifiedCollision) {
        logger.warn(
          { toolName: tool.name, serverId, collidingServer: unqualifiedCollision.serverId },
          'Tool name collision detected, only qualified names will work',
        );
      }

      // Store the tool in the registry
      this.toolRegistry.set(qualifiedName, {
        qualifiedName,
        definition: tool,
        serverId,
        client,
        transportType: client.transport as 'stdio' | 'http',
        reliability: {
          successCount: 0,
          failureCount: 0,
          circuitOpen: false,
        },
        performance: {
          avgResponseTimeMs: 0,
          callCount: 0,
          lastUsed: Date.now(),
        },
      });

      logger.debug({ qualifiedName }, 'Registered tool');
    }
  }

  /**
   * Resolves a tool name to a qualified name
   * @param toolName The tool name (can be qualified or unqualified)
   * @returns The qualified tool name
   * @throws {ToolNotFoundError} If the tool is not found
   */
  resolveToolName(toolName: string): string {
    // If already qualified, return as is
    if (toolName.includes(':')) {
      return toolName;
    }

    // Find tools with this unqualified name
    const matchingTools = [...this.toolRegistry.entries()].filter(([qualifiedName]) =>
      qualifiedName.endsWith(`:${toolName}`),
    );

    if (matchingTools.length === 0) {
      logger.error({ toolName }, 'Tool not found');
      throw new ToolNotFoundError(toolName);
    }

    if (matchingTools.length > 1) {
      logger.warn(
        {
          toolName,
          matches: matchingTools.map(([qualifiedName]) => qualifiedName),
        },
        'Multiple tools match unqualified name, using first match',
      );
    }

    return matchingTools[0][0]; // Return the first matching qualified name
  }

  /**
   * Gets a tool registry entry by name
   * @param toolName The tool name (can be qualified or unqualified)
   * @returns The tool registry entry or null if not found
   */
  getToolRegistryEntry(toolName: string): ToolRegistryEntry | null {
    try {
      const qualifiedName = this.resolveToolName(toolName);
      return this.toolRegistry.get(qualifiedName) || null;
    } catch (error) {
      return null;
    }
  }

  /**
   * Gets the tool description prompt for a server
   * @param serverId The server ID
   * @returns The tool description prompt or null if not found
   */
  getToolDescriptionPrompt(serverId: string): string | null {
    return this.toolDescriptionPrompts.get(serverId) || null;
  }

  /**
   * Gets all available tools
   * @returns An array of tool definitions
   */
  getAvailableTools(): ToolDefinition[] {
    return [...this.toolRegistry.values()].map((entry) => entry.definition);
  }

  /**
   * Gets all server tool description prompts
   * @returns A map of server IDs to tool description prompts
   */
  getAllToolDescriptionPrompts(): ServerToolDescriptionPrompts {
    return new Map(this.toolDescriptionPrompts);
  }

  /**
   * Validates tool arguments against the tool's JSON Schema
   * @param toolName The tool name (can be qualified or unqualified)
   * @param args The arguments to validate
   * @returns True if valid, or an object with errors if invalid
   * @throws {ToolNotFoundError} If the tool is not found
   */
  async validateToolArguments(
    toolName: string,
    args: Record<string, any>,
  ): Promise<boolean | { errors: any[] }> {
    const qualifiedName = this.resolveToolName(toolName);
    const toolEntry = this.toolRegistry.get(qualifiedName);

    if (!toolEntry) {
      throw new ToolNotFoundError(qualifiedName);
    }

    const schema = toolEntry.definition.parameters;

    // Skip validation if no schema is provided
    if (!schema) {
      return true;
    }

    // Compile the schema if needed
    const validate = this.ajv.compile(schema);
    const valid = validate(args);

    if (!valid) {
      return { errors: validate.errors || [] };
    }

    return true;
  }

  /**
   * Checks if the circuit breaker is open for a tool
   * @param toolName The tool name (can be qualified or unqualified)
   * @returns True if the circuit is open, false otherwise
   */
  isCircuitOpen(toolName: string): boolean {
    const entry = this.getToolRegistryEntry(toolName);
    if (!entry) {
      return false;
    }

    // Auto-reset circuit after cool-down period
    if (entry.reliability.circuitOpen && entry.reliability.lastFailure) {
      const timeSinceLastFailure = Date.now() - entry.reliability.lastFailure;
      if (timeSinceLastFailure > this.config.circuitResetTimeMs) {
        entry.reliability.circuitOpen = false;
        entry.reliability.failureCount = 0;
        logger.info(
          { toolName: entry.qualifiedName },
          'Circuit breaker auto-reset after cool-down period',
        );
      }
    }

    return entry.reliability.circuitOpen;
  }

  /**
   * Records a successful tool execution
   * @param toolName The tool name (can be qualified or unqualified)
   * @param executionTimeMs The execution time in milliseconds
   */
  recordToolSuccess(toolName: string, executionTimeMs: number): void {
    const entry = this.getToolRegistryEntry(toolName);
    if (!entry) {
      return;
    }

    // Update reliability - reduce failure count on success
    if (entry.reliability.failureCount > 0) {
      entry.reliability.failureCount--;
    }
    entry.reliability.successCount++;

    // Reset circuit breaker if it was open
    if (entry.reliability.circuitOpen) {
      entry.reliability.circuitOpen = false;
      logger.info(
        { toolName: entry.qualifiedName },
        'Circuit breaker closed after successful execution',
      );
    }

    // Update performance metrics
    entry.performance.callCount++;
    entry.performance.lastUsed = Date.now();

    // Exponential moving average for response time
    const alpha = 0.2; // Smoothing factor
    entry.performance.avgResponseTimeMs =
      alpha * executionTimeMs +
      (1 - alpha) * (entry.performance.avgResponseTimeMs || executionTimeMs);

    logger.debug(
      {
        toolName: entry.qualifiedName,
        executionTimeMs,
        avgResponseTimeMs: entry.performance.avgResponseTimeMs,
        successCount: entry.reliability.successCount,
      },
      'Recorded tool execution success',
    );
  }

  /**
   * Records a failed tool execution
   * @param toolName The tool name (can be qualified or unqualified)
   * @param error The error that occurred
   */
  recordToolFailure(toolName: string, error: Error): void {
    const entry = this.getToolRegistryEntry(toolName);
    if (!entry) {
      return;
    }

    // Update reliability metrics
    entry.reliability.failureCount++;
    entry.reliability.lastFailure = Date.now();

    // Check if circuit should open
    const shouldOpenCircuit = entry.reliability.failureCount >= this.config.circuitBreakerThreshold;

    if (shouldOpenCircuit && !entry.reliability.circuitOpen) {
      entry.reliability.circuitOpen = true;
      logger.warn(
        {
          toolName: entry.qualifiedName,
          failureCount: entry.reliability.failureCount,
          threshold: this.config.circuitBreakerThreshold,
        },
        'Circuit breaker opened due to consecutive failures',
      );
    }

    // Update performance metrics
    entry.performance.callCount++;
    entry.performance.lastUsed = Date.now();

    logger.debug(
      {
        toolName: entry.qualifiedName,
        failureCount: entry.reliability.failureCount,
        circuitOpen: entry.reliability.circuitOpen,
        error: error.message,
      },
      'Recorded tool execution failure',
    );
  }

  /**
   * Executes a tool
   * @param request The tool call request
   * @param options Execution options
   * @returns The tool call response
   * @throws {ToolNotFoundError} If the tool is not found
   * @throws {ToolValidationError} If the arguments are invalid
   * @throws {CircuitOpenError} If the circuit breaker is open
   * @throws {ToolExecutionError} If the tool execution fails
   */
  async executeTool(
    request: ToolCallRequest,
    options: ToolExecutionOptions = {},
  ): Promise<ToolCallResponse> {
    const startTime = Date.now();
    const qualifiedToolName = this.resolveToolName(request.toolName);

    logger.info(
      {
        toolName: qualifiedToolName,
        requestId: request.requestId,
      },
      'Executing tool',
    );

    const toolEntry = this.toolRegistry.get(qualifiedToolName);

    if (!toolEntry) {
      throw new ToolNotFoundError(qualifiedToolName);
    }

    // Check circuit breaker
    if (this.isCircuitOpen(qualifiedToolName)) {
      const error = new CircuitOpenError(
        qualifiedToolName,
        toolEntry.reliability.failureCount,
        toolEntry.reliability.lastFailure!,
      );
      logger.warn(
        {
          toolName: qualifiedToolName,
          failureCount: toolEntry.reliability.failureCount,
        },
        'Tool execution blocked by circuit breaker',
      );
      throw error;
    }

    // Validate arguments
    const validationResult = await this.validateToolArguments(qualifiedToolName, request.arguments);
    if (validationResult !== true) {
      const error = new ToolValidationError(qualifiedToolName, (validationResult as any).errors);
      logger.warn(
        {
          toolName: qualifiedToolName,
          errors: (validationResult as any).errors,
          arguments: request.arguments,
        },
        'Tool arguments validation failed',
      );
      throw error;
    }

    // Setup execution with timeout
    const abortController = new AbortController();
    const timeoutMs =
      request.timeoutMs || toolEntry.timeoutMs || options.timeoutMs || this.config.defaultTimeoutMs;

    const timeout = setTimeout(() => {
      abortController.abort();
    }, timeoutMs);

    this.activeExecutions.set(request.requestId, { abortController, timeout });

    try {
      logger.debug(
        {
          toolName: qualifiedToolName,
          requestId: request.requestId,
          arguments: request.arguments,
          timeoutMs,
        },
        'Calling tool with arguments',
      );

      // Execute the tool
      const result = await toolEntry.client.callTool(toolEntry.definition.name, request.arguments, {
        signal: abortController.signal,
      });

      const executionTimeMs = Date.now() - startTime;

      // Record success
      this.recordToolSuccess(qualifiedToolName, executionTimeMs);

      logger.info(
        {
          toolName: qualifiedToolName,
          requestId: request.requestId,
          executionTimeMs,
        },
        'Tool execution completed successfully',
      );

      return {
        toolName: qualifiedToolName,
        result,
        requestId: request.requestId,
        executionTimeMs,
      };
    } catch (error) {
      const executionTimeMs = Date.now() - startTime;

      // Record failure
      this.recordToolFailure(qualifiedToolName, error as Error);

      logger.error(
        {
          toolName: qualifiedToolName,
          requestId: request.requestId,
          error,
          executionTimeMs,
        },
        'Tool execution failed',
      );

      throw new ToolExecutionError(`Failed to execute tool: ${qualifiedToolName}`, error as Error, {
        toolName: qualifiedToolName,
        serverId: toolEntry.serverId,
        requestId: request.requestId,
        arguments: request.arguments,
      });
    } finally {
      clearTimeout(timeout);
      this.activeExecutions.delete(request.requestId);
    }
  }

  /**
   * Aborts a running tool execution
   * @param requestId The request ID
   * @returns True if aborted, false if not found
   */
  async abortToolExecution(requestId: string): Promise<boolean> {
    const execution = this.activeExecutions.get(requestId);

    if (!execution) {
      return false;
    }

    clearTimeout(execution.timeout);
    execution.abortController.abort();
    this.activeExecutions.delete(requestId);

    logger.info({ requestId }, 'Aborted tool execution');

    return true;
  }

  /**
   * Shuts down the MCP Coordinator
   * Disconnects all clients and cleans up resources
   */
  async shutdown(): Promise<void> {
    logger.info('Shutting down MCP Coordinator');

    // Abort any active executions
    for (const [requestId, execution] of this.activeExecutions.entries()) {
      clearTimeout(execution.timeout);
      execution.abortController.abort();
      logger.debug({ requestId }, 'Aborted active execution during shutdown');
    }
    this.activeExecutions.clear();

    // Disconnect all clients
    for (const [serverId, client] of this.clients.entries()) {
      try {
        await client.disconnect();
        logger.debug({ serverId }, 'Disconnected MCP client');
      } catch (error) {
        logger.warn({ serverId, error }, 'Failed to disconnect MCP client');
      }
    }

    this.clients.clear();
    this.toolRegistry.clear();
    this.toolDescriptionPrompts.clear();

    logger.info('MCP Coordinator shutdown complete');
  }
}

File: src/core/mcp_coordinator.py
# src/core/mcpCoordinator.py

from dataclasses import dataclass
from typing import Dict, List, Optional, Any, Tuple
import json
import asyncio
import sys
import traceback
from pathlib import Path

from mcp import ClientSession, StdioServerParameters, types, Tool
from mcp.client.stdio import stdio_client

@dataclass
class ServerConfig:
    """Configuration for an MCP server."""
    name: str
    description: str
    type: str
    transport: str
    command: Optional[str] = None
    args: Optional[List[str]] = None
    url: Optional[str] = None
    env: Optional[Dict[str, str]] = None
    timeout_ms: int = 10000

@dataclass
class ToolRegistryEntry:
    """Entry in the tool registry."""
    qualified_name: str
    definition: types.Tool
    server_id: str
    client: ClientSession # Note: This client lives in a background task
    transport_type: str
    reliability: Dict[str, Any]
    performance: Dict[str, Any]

class MCPCoordinator:
    def __init__(self, config_path: str):
        """Initialize the MCP Coordinator."""
        self.config_path = Path(config_path)
        self.tool_registry: Dict[str, ToolRegistryEntry] = {}
        self.clients: Dict[str, ClientSession] = {} # Stores active client sessions
        # Note: _client_contexts is removed as contexts are managed within tasks now
        self._server_tasks: Dict[str, asyncio.Task] = {}
        self._tg: Optional[asyncio.TaskGroup] = None
        self._shutdown_event = asyncio.Event()
        # Circuit breaker settings (kept for potential future use)
        self.circuit_threshold = 5
        self.circuit_reset_time_ms = 60000

    async def _load_config(self) -> Dict[str, ServerConfig]:
        """Load and validate the mcp.json configuration."""
        print(f"Coordinator: Loading config from {self.config_path}...")
        try:
            with open(self.config_path) as f:
                raw_config = json.load(f)
            configs = {}
            for server_id, server_data in raw_config.get("servers", {}).items():
                server_data.pop('id', None)
                configs[server_id] = ServerConfig(**server_data)
            print(f"Coordinator: Config loaded for {len(configs)} servers.")
            return configs
        except Exception as e:
            print(f"Coordinator ERROR: Failed to load config: {e}")
            raise ValueError(f"Failed to load MCP configuration from {self.config_path}: {e}")

    async def _discover_tools_for_client(self, server_id: str, client: ClientSession, server_config: ServerConfig):
        """Discover and register tools for a specific client."""
        try:
            tools_result = await client.list_tools()
            tools = tools_result.tools if hasattr(tools_result, 'tools') else []
            count = 0
            for tool in tools:
                qualified_name = f"{server_id}:{tool.name}"
                self.tool_registry[qualified_name] = ToolRegistryEntry(
                    qualified_name=qualified_name,
                    definition=tool,
                    server_id=server_id,
                    client=client, # The client session managed by this task
                    transport_type=server_config.transport,
                    reliability={"success_count": 0, "failure_count": 0, "circuit_open": False, "last_failure": None},
                    performance={"avg_response_time_ms": 0, "call_count": 0, "last_used": None}
                )
                count += 1
            if count > 0:
                print(f"  - [{server_id}] Registered {count} tools.")
        except Exception as e:
            print(f"  - [{server_id}] ERROR during tool discovery: {e}")

    async def _manage_server_connection(self, server_id: str, server_config: ServerConfig, setup_event: asyncio.Event):
        """Dedicated task managing a single server connection lifecycle."""
        client_session_ref = None
        task_failed = False
        try:
            if server_config.transport == "stdio":
                if not server_config.command:
                    raise ValueError(f"Missing command for stdio server {server_id}")

                server_params = StdioServerParameters(
                    command=server_config.command,
                    args=server_config.args or [],
                    env=server_config.env or {},
                )

                async with stdio_client(server_params) as (read, write):
                    async with ClientSession(read, write) as session:
                        client_session_ref = session
                        await session.initialize()
                        self.clients[server_id] = session
                        await self._discover_tools_for_client(server_id, session, server_config)
                        setup_event.set()
                        await self._shutdown_event.wait()

            elif server_config.transport == "sse":
                print(f"[{server_id}] Warning: Unsupported transport 'sse'.")
                task_failed = True
                setup_event.set()
            
            else:
                 print(f"[{server_id}] Warning: Unknown transport '{server_config.transport}'.")
                 task_failed = True
                 setup_event.set()

        except asyncio.CancelledError:
            pass
        except Exception as e:
            print(f"[{server_id}] ERROR in server task: {e}")
            task_failed = True
            if not setup_event.is_set():
                setup_event.set()
        finally:
            if server_id in self.clients and self.clients.get(server_id) is client_session_ref:
                del self.clients[server_id]
            keys_to_delete = [key for key in self.tool_registry if key.startswith(f"{server_id}:")]
            for key in keys_to_delete:
                if key in self.tool_registry:
                     del self.tool_registry[key]

    async def initialize(self) -> None:
        """Initialize by launching server tasks and waiting for their setup."""
        print("Coordinator: Initializing...")
        config = await self._load_config()

        if not config:
            print("Coordinator: No servers defined.")
            return

        self._shutdown_event.clear()
        self._server_tasks.clear()
        setup_events: Dict[str, asyncio.Event] = {}
        wait_tasks_map: Dict[asyncio.Task, str] = {}

        if self._tg is None:
            raise RuntimeError("TaskGroup not initialized. MCPCoordinator must be used with 'async with'.")

        print(f"Coordinator: Launching {len(config)} server tasks...")
        for server_id, server_config in config.items():
            setup_event = asyncio.Event()
            setup_events[server_id] = setup_event
            server_task = self._tg.create_task(
                self._manage_server_connection(server_id, server_config, setup_event)
            )
            self._server_tasks[server_id] = server_task
            wait_task = asyncio.create_task(setup_event.wait(), name=f"wait_setup_{server_id}")
            wait_tasks_map[wait_task] = server_id

        setup_timeout = 30.0
        
        if not wait_tasks_map:
             return

        done, pending = await asyncio.wait(wait_tasks_map.keys(), timeout=setup_timeout)

        failed_servers = []
        successful_servers = []

        if pending:
            print(f"Coordinator WARNING: Server setup timed out for:")
            for wait_task in pending:
                server_id = wait_tasks_map.get(wait_task, "unknown_timeout")
                print(f"  - {server_id}")
                wait_task.cancel()
                failed_servers.append(server_id)
                if server_id != "unknown_timeout" and server_id in self._server_tasks:
                    self._server_tasks[server_id].cancel()
            
        for wait_task in done:
             server_id = wait_tasks_map.get(wait_task, "unknown_done")
             if wait_task.cancelled():
                 if server_id not in failed_servers: failed_servers.append(server_id)
             elif wait_task.exception():
                  exc = wait_task.exception()
                  print(f"Coordinator ERROR: Waiting for {server_id} setup failed: {exc}")
                  if server_id not in failed_servers: failed_servers.append(server_id)
                  if server_id in self._server_tasks: self._server_tasks[server_id].cancel()
             else:
                  successful_servers.append(server_id)

        print("\n--- Coordinator Initialization Summary ---")
        if successful_servers:
            print("Successfully Initialized Servers:")
            for server_id in sorted(successful_servers):
                 print(f"  - {server_id}")
        if failed_servers:
            print("Failed/Timeout Servers:")
            for server_id in sorted(failed_servers):
                 print(f"  - {server_id}")
        if not successful_servers and not failed_servers:
             print("No servers were processed.")
        print("----------------------------------------")

    async def call_tool(self, qualified_tool_name: str, arguments: Dict[str, Any]) -> Any:
        """Calls a registered tool by its qualified name."""
        if qualified_tool_name not in self.tool_registry:
            raise ValueError(f"Tool '{qualified_tool_name}' not found in registry.")
        
        tool_entry = self.tool_registry[qualified_tool_name]
        client = tool_entry.client
        tool_name = tool_entry.definition.name
        
        try:
            result = await client.call_tool(tool_name, arguments=arguments)
            return result
        except Exception as e:
            print(f"Coordinator ERROR: Calling tool '{qualified_tool_name}' failed: {e}")
            raise

    # --- Context Manager Implementation ---
    async def __aenter__(self):
        """Enter the coordinator context, creating task group and initializing."""
        self._tg = asyncio.TaskGroup()
        await self._tg.__aenter__()
        try:
            await self.initialize()
        except Exception as init_exc:
             print(f"Coordinator ERROR: Initialization failed: {init_exc}. Exiting context early.")
             self._shutdown_event.set()
             await self._tg.__aexit__(type(init_exc), init_exc, init_exc.__traceback__)
             self._tg = None
             raise
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Exit the coordinator context, cleaning up resources."""
        if self._tg is None:
            return

        self._shutdown_event.set()

        exit_result = False
        try:
            exit_result = await self._tg.__aexit__(exc_type, exc_val, exc_tb)
        except Exception as tg_exit_exc:
             print(f"Coordinator ERROR: Exception during TaskGroup exit: {tg_exit_exc}")

        self.clients.clear()
        self.tool_registry.clear()
        self._server_tasks.clear()
        self._tg = None

            